### PR TITLE
Package plugin_extras

### LINKED ISSUE #4147: Plugin Extras
Extending #3072 to implement `plugin_extras` fields on:
* Users
* Groups and Organizations (same underlying model)
* Datasets
* Resources

`plugin_extras` is a JSON field that is not exposed through the user interface or API (by default.) It's intended as a generic way for plugins to attach data to Users, Groups, Organizations, Datasets and Resources. 

By convention each plugin will store its own data in a separate JSON object, e.g. `my_user.plugin_extras['funky_auth']` would be the place a `ckanext.funky_auth` extension could store its data related to `my_user`.

For API-level migration of data from one CKAN instance to another, sysadmin users may request `plugin_extras` data by passing `include_plugin_extras=true` to the appropriate `*_show` API, and `plugin_extras` values passed to `*_update` and `*_create` APIs will be accepted when called by a sysadmin user. By default (and for all other users) this key is ignored and not returned.

Plugin extras are **not** revisioned and **not** copied into activity records. Values stored are private and hidden to normal users.

Plugins may choose to expose values they store in `plugin_extras` and/or allow them to be updated from the API and front end by chaining actions or customizing validation rules, **not** through any new interface introduced by this change.

This PR only creates the `plugin_extras` fields to enable future plugins to do something actually useful, e.g. https://github.com/ckan/ideas-and-roadmap/issues/211

### PR DESCRIPTION
**Pull request by TomeCirun**
Fixes #4147

### Features:

- [ ] includes tests covering changes
- [ ] includes updated documentation
- [ ] includes user-visible changes
- [ ] includes API changes
- [ ] includes bugfix for possible backport

Please [X] all the boxes above that apply

### CODE REVIEW

**Comment by TomeCirun** [comment]
@ward, I would need some help here, as you know in package_show, use_cache by default is True, therefore I'm struggling to get the plugin_extras from solar, I would need help with solar,  please let me know if I'm on the right track and what action should I take next.

p.s Should I delete package_extra?

Thanks

**Comment by tino097** [comment]
Ian's commen to continue with the [discsussion](https://github.com/ckan/ckan/issues/4147#issuecomment-1158014489)

**Comment by TomeCirun** [comment]
Hey @wardi, is there anything else that can be done in this PR?

**Comment by wardi** [comment]
We need to look at how plugin_data interacts with the search index. It should be possible to copy fields from plugin_data into solr fields from `before_index` (e.g. for a plugin that adds search terms from file contents or a plugin that uses ML to improve search results) but plugin_data itself shouldn't be stored in the `validated_data_dict` or `data_dict` values by default.

File: ckan/logic/action/get.py [hunk 1]
```diff
@@ -1009,6 +1012,7 @@ def package_show(context: Context, data_dict: DataDict) -> ActionResult.PackageS
     package_dict = None
     use_cache = (context.get('use_cache', True))
     package_dict_validated = False
+    include_plugin_data = asbool(data_dict.get('include_plugin_data', False))
 
     if use_cache:
```
[the remark this answers is not in the document]
Comment (TomeCirun) [reply ckan/logic/action/get.py:1017]: @wardi Yes, plugin_data won't be in the cache, it's only available if the requester is sysadmin and he must pass include_plugin_data=True to package_show alongside use_cache=False

[the remark this answers is not in the document]
Comment (TomeCirun) [reply ckan/logic/action/get.py:1017]: Yep, you are right, that makes sense, thanks for the help

**Comment by TomeCirun** [comment]
> We need to look at how plugin_data interacts with the search index. It should be possible to copy fields from plugin_data into solr fields from `before_index` (e.g. for a plugin that adds search terms from file contents or a plugin that uses ML to improve search results) but plugin_data itself shouldn't be stored in the `validated_data_dict` or `data_dict` values by default.


How should I achieve that? through the tests maybe? I have no idea how before_dataset_index works, maybe you can share some examples of it so I can get the idea? 
Thanks

**Comment by wardi** [comment]
> How should I achieve that? through the tests maybe? I have no idea how before_dataset_index works, maybe you can share some examples of it so I can get the idea? @wardi Thanks

This could be future work separate from this PR, but would you check to make sure that `plugin_data` *doesn't* appear in the data stored in `validated_data_dict` https://github.com/ckan/ckan/blob/0a596b8394dbf9582902853ad91450d2c0d7959b/ckan/lib/search/index.py#L127 ?

**Comment by TomeCirun** [comment]
I sent the data from the tests, plugin_data doesn't appear in the data stored in  validated_data_dict, but i m getting this error:

pysolr.SolrError: Solr responded with an error (HTTP 400): [Reason: Error:[doc=fe29f7f546eee02a6fed907051abe1e1]  Unknown operation for the an atomic update: plugin1]

from this [line](https://github.com/ckan/ckan/blob/0a596b8394dbf9582902853ad91450d2c0d7959b/ckan/lib/search/index.py#L308)

**Comment by TomeCirun** [comment]
@wardi it seems the error comes from the nested dict, I also encountered the same error while using `ckanext-scheming` on CKAN 2.10

image

**Comment by TomeCirun** [comment]
I opened an issue in their repo django-haystack/pysolr#385

**Comment by TomeCirun** [comment]
@wardi polite reminder

File: ckan/tests/logic/action/test_create.py [hunk 2]
```diff
@@ -1946,3 +1946,60 @@ def test_member_create_raises_if_object_type_is_invalid(self):
...
+            }
+        }
+        context = {
+            "user": user["name"],
+            "ignore_auth": False,
+        }
+        created_pkg = helpers.call_action(
+            'package_create', context=context, **pkg_dict
+        )
+        assert "plugin_data" not in created_pkg
```
Comment (wardi) [inline ckan/tests/logic/action/test_create.py:2003]: better to use the same sql as above to check for the existence of plugin_data

**Comment by wardi** [comment]
`plugin_extras` appears in the migration filename, a comment and a couple variables. Would you update those names to `plugin_data` and make the small change to the test mentioned above?

**Comment by TomeCirun** [comment]
> `plugin_extras` appears in the migration filename, a comment and a couple variables. Would you update those names to `plugin_data` and make the small change to the test mentioned above?


@wardi done :)
Thanks

File: ckan/tests/logic/action/test_create.py [hunk 3]
```diff
@@ -2001,5 +2001,9 @@ def test_ignored_on_create_if_non_sysadmin(self):
             'package_create', context=context, **pkg_dict
         )
         assert "plugin_data" not in created_pkg
-        pkg = model.Package.get(created_pkg["id"])
-        assert pkg.plugin_data is None
+
+        plugin_daya_from_db = model.Session.execute(
```
Comment (wardi) [inline ckan/tests/logic/action/test_create.py:2005]: daya?

Comment (TomeCirun) [reply ckan/tests/logic/action/test_create.py]: It's a typo, tnx, good catch btw the failed test is not related to my changes

**Pull request merged**